File: task4_solution.kt
import java.util.Scanner

data class Reaction(val from:HashSet<String>, val to: HashSet<String>)

fun main() {
    val sc = Scanner(System.`in`)
    val compounds = HashSet<String>()
    compounds.addAll(sc.nextLine().split(" "))
    val reactions = ArrayList<Reaction>()
    while (sc.hasNextLine()) {
        val array = sc.nextLine().split("=")
        val r = Reaction(HashSet(array[0].split(" ")), HashSet(array[1].split(" ")))
        reactions.add(r)
    }
    var newCompoundAdded: Boolean

    do {
        newCompoundAdded = false
        val iterator = reactions.iterator()
        for (r in iterator) {
            if (compounds.containsAll(r.from)) {
                compounds.addAll(r.to)
                iterator.remove()
                newCompoundAdded = true
                //println(compounds)
            }
        }
    } while (newCompoundAdded)
    println(compounds.sorted().joinToString(" ") { it })

}

File: task4_solution.py
to_use = set(input().split())

reactions = []

while True:
    try:
        data = input().split('=')
    except Exception:
        break
    data0 = data[0].split(' ')
    data1 = data[1].split(' ')
    reactions.append([data0, data1])

prev = 0

while len(to_use) != prev:
    prev = len(to_use)
    for i in range(len(reactions)):
        for chem in reactions[i][0]:
            if chem not in to_use:
                break
        else:
            for chem in reactions[i][1]:
                to_use.add(chem)

to_use = sorted(to_use)

for chem in to_use:
    print(chem, end=" ")
